core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,401 @@
+      Thread.sleep(5 * 1000L); 1
+    }).count()
+    assert(jobResult === 1)
+  }
+
+  test("decommission workers ensure that shuffle output is regenerated even with shuffle service") {
+    val conf = appConf
+    conf.set(config.Tests.TEST_NO_STAGE_RETRY, true)
+    conf.set(config.SHUFFLE_MANAGER, "sort")
+    conf.set(config.SHUFFLE_SERVICE_ENABLED, true)
+    makeWorkers(2)
+    val ss = new ExternalShuffleServiceHolder(conf)
+    sc = createSparkContext(conf)
+
+    // Here we will create a 2 stage job: The first stage will have two tasks and the second stage
+    // will have one task. The two tasks in the first stage will be long and short. We decommission
+    // and kill the worker after the short task is done. Eventually the driver should get the
+    // executor lost signal for the short task executor. This should trigger regenerating
+    // the shuffle output since we cleanly decommissioned the executor, despite running with an
+    // external shuffle service.
+    try {
+      val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+      val task0Killed = new AtomicBoolean(false)
+      // single task job
+      val listener = new RootStageAwareListener {
+        override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+          val taskInfo = taskEnd.taskInfo
+          assert(taskInfo.index <= 1, s"Unknown task index ${taskInfo.index}")
+          if (taskInfo.index == 0) {
+            if (task0Killed.compareAndSet(false, true)) {
+              assert(taskInfo.successful)
+              // Since this task hasn't been killed before, it should still be at its first attempt.
+              assert(taskInfo.attemptNumber === 0)
+              val workerInfo = executorIdToWorkerInfo(taskInfo.executorId)
+              decommissionWorkerOnMaster(workerInfo, "Kill early done map worker")
+              killWorkerAfterTimeout(workerInfo, 0)
+              logInfo(s"Killed the node ${workerInfo.hostPort} that was running the early task")
+            } else {
+              assert(taskInfo.successful)
+              // The first attempt of this task should have failed since it was killed. Thus,
+              // either the task attempt or the stage attempt number should be more than 0.
+              assert(taskInfo.attemptNumber > 0 || taskEnd.stageAttemptId > 0)
+            }
+          } else {
+            // The second task is never touched and thus should succeed on the first attempt.
+            assert(taskInfo.successful)
+            assert(taskInfo.attemptNumber === 0)
+          }
+        }
+      }
+      sc.addSparkListener(listener)
+      val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex((pid, iter) => {
+        val sleepTimeSeconds = if (pid == 0) 1 else 10
+        Thread.sleep(sleepTimeSeconds * 1000L)
+        List.fill(pid + 1)(pid * 2 + 1).iterator
+      }, preservesPartitioning = true).repartition(1).sum()
+      assert(jobResult > 1)
```

Comment:
Do you expect here a result changing by different runs? i mean why not an equality with a constant is given?  
If there is a reason for this please mention it in a comment above.